OpenRA.Mods.Common/Traits/World/PathCacheStorage.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class PathCacheStorage : ICacheStorage<List<CPos>>
+	{
+		class CachedPath
+		{
+			public List<CPos> Result;
+			public int Tick;
+		}
+
+		private const int MaxPathAge = 50;
+		private readonly IWorld world;
+		Dictionary<string, CachedPath> cachedPaths = new Dictionary<string, CachedPath>();
+
+		public PathCacheStorage(IWorld world)
+		{
+			this.world = world;
+		}
+
+		public void Remove(string key)
+		{
+			cachedPaths.Remove(key);
+		}
+
+		public void Store(string key, List<CPos> data)
+		{
+			foreach (var cachedPath in cachedPaths.Where(p => world.WorldTick - p.Value.Tick > MaxPathAge).ToList())
```

Comment:
I think there's (a little) room for improvement here. I don't really know how costly those linq queries are in the great scheme of thing, but even in a moderately busy game, you can already have ~100 cached paths, and you need to walk the whole thing every time. Do you think it would make sense to keep track of the tick that the last cleanup took place, so that you don't do it multiple times in the same tick?

It's probably premature optimization, but I figure the busy games could need all the perf they can get.